src/content/docs/apm/agents/python-agent/python-agent-api/setllmtokencountcallback-python-agent-api.mdx

```diff
@@ -0,0 +1,160 @@
+If you would still like to capture token counts for Llm events in this instance, you can implement a callback in your app code to determine the token counts locally and send this information to New Relic.
+In most cases, this API will be called exactly once soon after the agent has initialized. However, you may make multiple calls to this API if you would like.
+Each fresh call made to the endpoint will overwrite the previously registered callback with the new one that is provided. To unset the callback completely, `None` can be passed in place of the original callback.
+
+## API Parameters
+
+<table>
+  <thead>
+    <tr>
+      <th width="25%">
+        Parameter
+      </th>
+
+      <th>
+        Description
+      </th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <tr>
+      <td>
+        `callback`
+
+        _callable_ or _None_
+      </td>
+
+      <td>
+        Required. The callback to calculate token counts. To unset the current callback, pass `None` in instead of a callback function.
+      </td>
+    </tr>
+
+    <tr>
+      <td>
+        `application`
+
+        _object_
+      </td>
+
+      <td>
+        Optional. The specific application object to associate the API call with. An application object can be obtained using the [`newrelic.agent.application`](/docs/apm/agents/python-agent/python-agent-api/application-python-agent-api/) function.
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+## Return values
+
+None
+
+## Callback Requirements
+
+Provided callbacks must return a positive integer token count value for a given piece of content or no token count will be captured on Llm events.
+
+## Callback Parameters
+
+<table>
+  <thead>
+    <tr>
+      <th width="25%">
+        Parameter
+      </th>
+
+      <th>
+        Description
+      </th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <tr>
+      <td>
+        `model`
+
+        _string_
+      </td>
+
+      <td>
+        Required. The name of the LLM model.
+      </td>
+    </tr>
+
+    <tr>
+      <td>
+        `content`
+
+        _string_
+      </td>
+
+      <td>
+        Required. The message content/ prompt or embedding input.
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+## Examples
+
+### Calculating token counts and registering the callback
+
+Example with tiktoken:
+
+```py
+import newrelic.agent
+def token_count_callback(model, content):
+    """
+    Calculate token counts locally based on the model being used and the content.
+    This callback will be invoked for each message as part of a communication with the LLM.
+    If the application supports more than one model, it may require finding libraries for
+    each model to support token counts appropriately.
+
+    Arguments:
+    model -- name of the model provided by or returned from the LLM
+    content -- context, input or output of the LLM
+    """
+    if "gpt" in model:
+        import tiktoken
+
+        enc = tiktoken.encoding_for_model(model)
```

Comment:
This raises a KeyError if the model is not found so we should probably include handling that in this example so customers can just copy-paste this into their code. Since we record a warning for <0 values it would be better to return None in Python's API instead of 0 as well. Instead of the if statement and all this stuff let's just simplify this to the following callback implementation:

```
import tiktoken

try:
    enc = tiktoken.encoding_for_model(model)
except KeyError:
    return None  # Unknown model.
return len(enc.encode(content))
```